lgc/docs/LgcOverview.md

```diff
@@ -126,6 +126,102 @@ The front-end flow is:
+such as the offset of a descriptor table in user data, and the offset of a descriptor within its
+descriptor table. The LGC ELF linker resolves such relocs with reference to the pipeline state.
+In addition, if vertex layout is not available when compiling the VS, then the LGC ELF linker glues
+a "fetch shader" on to the front, and, similarly, if color export information is not available when
+compiling the FS, then the LGC ELF linker glues an "export shader" on to the end. Compilation of each
+shader places metadata into the PAL metadata .note record for the linker to generate these glue shaders.
+
+This scheme has some restrictions on what the pipeline is allowed to do. Most notably, if resource mapping
+(the layout of user data and descriptor tables) is not available at shader compile time, then it is not
+allowed to have a descriptor in the root table instead of in the descriptor table for its set. That means
+that XGL "dynamic descriptors" are not allowed. Also, it seems like some built-in inputs in the FS
+would not work, if they are implemented as generic inputs that need to be exported by the VS.
+
+#### Old partial pipeline compilation scheme
+
+LGC has a hook part way through compilation, just after determining inter-shader output/input packing,
+in which it calls a callback function (provided by the front-end as the `checkShaderCacheFunc` arg to
+`Pipeline::generate` method), providing a hash of input/output usage and mapping for each shader stage,
+and allowing the callback to clear bits in the shader stage mask to disable the continued compilation of
+part of the pipeline.
+
+The LLPC front-end uses this to implement a partial pipeline compilation scheme. Considering the two
+parts of a graphics pipeline (the non-FS part (VS,TCS,TES,GS) and the FS), it uses the hashes, in
+conjunction with its own hashes of the shader stages, to determine whether it has compiled the same
+partial pipeline with the same input/output usage before. The ELF result of compilation is added to
+its cache for future compiles to find.
+
+To do this, the LLPC front-end has to extract the partial pipeline it wants from a pipeline ELF, then
+link two partial pipelines together. It does this with its own ELF extracting and linking code, not
+using the LGC ELF linker.
+
+#### New part-pipeline compilation scheme
+
+LGC supports another scheme, similar in concept but not needing compilation of the whole pipeline
+to proceed to a cache-checking point, and allowing use of the LGC ELF linker.
+
+This scheme is based on the observation that you can compile the FS without reference to other shaders
+(but with the pipeline state), and it can still pack its inputs (i.e. not reserve an attribute in the
+attribute space if it never uses it), as long as the compile of the other part of the pipeline gets
+the information about that packing. The FS input information also includes where the FS uses a built-in
+that needs to be exported from the VS.
```

Comment:
Is this because in-out packing works backwards, from outputs to inputs?

Reply:
Yes; clarified.